File: Ben.iOS/BenKit/Sources/BenKit/Stores/SessionStore.swift
import Foundation
import Observation

/// The auth state machine the whole app hangs off:
///
///     signedOut → authenticating → (twoFactorChallenge →) fetchingIdentity → signedIn
///
/// Session-ended (a failed refresh, or a 401 on a live token) is an INTERRUPT:
/// it lands back in `signedOut` with a banner, and anonymous surfaces keep
/// working — never a sign-in wall.
@Observable
@MainActor
public final class SessionStore {
    public enum State: Equatable {
        case signedOut
        case authenticating
        /// Password was right; a code is needed. Credentials are held in
        /// MEMORY ONLY for the retry, never persisted.
        case twoFactorChallenge
        case fetchingIdentity
        case signedIn(MeResponse)
    }

    public private(set) var state: State = .signedOut
    /// A human sentence for the sign-in form (wrong password, server prose).
    public private(set) var errorMessage: String?
    /// Non-nil while the auth endpoint has us in a 429 window.
    public private(set) var retryAfter: TimeInterval?
    /// Set when the session ended out from under the user; cleared on sign-in.
    public private(set) var sessionEndedBanner = false

    private let auth: IdentityAuthClient
    private let tokens: TokenSession
    private let api: APIClient
    private let identity: IdentityStorage

    /// Signed in from the identity kept on the phone, because the server couldn't be reached at launch. The next
    /// `restore()` asks the server again.
    public private(set) var identityIsKept = false

    // Held only between the password step and the 2FA retry.
    private var pendingEmail: String?
    private var pendingPassword: String?
    private var eventTask: Task<Void, Never>?
    /// Set around a deliberate sign-out so the session-ended event it emits
    /// (delivered async on the stream) doesn't raise the interrupt banner.
    private var expectingDeliberateEnd = false

    public var me: MeResponse? {
        if case .signedIn(let me) = state { return me }
        return nil
    }

    public init(auth: IdentityAuthClient, tokens: TokenSession, api: APIClient, identity: IdentityStorage = InMemoryIdentityStorage()) {
        self.auth = auth
        self.tokens = tokens
        self.api = api
        self.identity = identity
        // Surface refresh failures from anywhere in the app as the interrupt.
        eventTask = Task { [weak self] in
            let stream = await tokens.events()
            for await event in stream {
                guard let self else { return }
                if case .sessionEnded = event { self.handleSessionEnded() }
            }
        }
    }

    /// Cold start: tokens in the Keychain mean optimistic sign-in pending
    /// `api/me`. A stale token (reinstall, revocation) lands QUIETLY in
    /// signed-out — no error dialog on first launch.
    ///
    /// With no signal, a person whose identity is kept on the phone is signed in as that person, and a later
    /// `restore()` (the app coming back to the foreground) confirms it with the server.
    public func restore() async {
        guard await tokens.isSignedIn else { return }
        if identityIsKept, case .signedIn = state {
            await fetchIdentity(quietOnFailure: true)
            return
        }
        guard state == .signedOut else { return }
        state = .fetchingIdentity
        await fetchIdentity(quietOnFailure: true)
    }

    public func signIn(email: String, password: String) async {
        guard state == .signedOut || state == .twoFactorChallenge else { return }
        errorMessage = nil
        retryAfter = nil
        state = .authenticating
        await attempt(LoginRequest(email: email, password: password),
                      rememberFor2FA: (email, password))
    }

    /// The 2FA retry: same `/login` call, same credentials, plus exactly one
    /// of the two code fields.
    public func submitTwoFactor(code: String, isRecoveryCode: Bool) async {
        guard state == .twoFactorChallenge,
              let email = pendingEmail, let password = pendingPassword else { return }
        errorMessage = nil
        state = .authenticating
        let trimmed = code.replacingOccurrences(of: " ", with: "")
            .replacingOccurrences(of: "-", with: "")
        let request = LoginRequest(
            email: email, password: password,
            twoFactorCode: isRecoveryCode ? nil : trimmed,
            twoFactorRecoveryCode: isRecoveryCode ? trimmed : nil)
        await attempt(request, rememberFor2FA: (email, password))
    }

    public func cancelTwoFactor() {
        guard state == .twoFactorChallenge || state == .authenticating else { return }
        clearPending()
        state = .signedOut
        errorMessage = nil
    }

    /// Called when somebody signs out on purpose — not when a session merely expires — so what the app keeps on the
    /// phone for them (saved event passes) goes before the next person picks the phone up.
    public var onDeliberateSignOut: (@MainActor () -> Void)?

    public func signOut() async {
        onDeliberateSignOut?()
        clearPending()
        // Signing out on purpose is not an interrupt — the event this emits
        // must not raise the banner.
        expectingDeliberateEnd = true
        await tokens.endSession()
        forgetIdentity()
        sessionEndedBanner = false
        state = .signedOut
        errorMessage = nil
    }

    public func dismissSessionEndedBanner() {
        sessionEndedBanner = false
    }

    // MARK: - Internals

    private func attempt(_ request: LoginRequest, rememberFor2FA: (String, String)) async {
        switch await auth.login(request) {
        case .success(let response):
            clearPending()
            await tokens.adopt(response)
            sessionEndedBanner = false
            state = .fetchingIdentity
            await fetchIdentity(quietOnFailure: false)
        case .requiresTwoFactor:
            (pendingEmail, pendingPassword) = rememberFor2FA
            // A wrong 2FA code comes back as requiresTwoFactor again; say so
            // when the user has already been on this screen.
            if request.twoFactorCode != nil || request.twoFactorRecoveryCode != nil {
                errorMessage = "That code didn't work — try the current one."
            }
            state = .twoFactorChallenge
        case .invalidCredentials:
            clearPending()
            errorMessage = "Invalid email or password."
            state = .signedOut
        case .rateLimited(let after):
            clearPending()
            retryAfter = after ?? 60
            errorMessage = nil
            state = .signedOut
        case .failed(let reason):
            clearPending()
            errorMessage = reason ?? "The server couldn't be reached."
            state = .signedOut
        }
    }

    /// Adopts a session that some OTHER door established — today, Sign in with Apple, which
    /// puts the credentials in the token session itself and never sees a password. The state
    /// machine still has to catch up, or the app stays on its sign-in sheet holding a valid token.
    public func adoptExternalSignIn() async {
        clearPending()
        errorMessage = nil
        retryAfter = nil
        sessionEndedBanner = false
        state = .fetchingIdentity
        await fetchIdentity(quietOnFailure: false)
    }

    private func fetchIdentity(quietOnFailure: Bool) async {
        let result = await api.load(Endpoint(.get, "api/me"), as: MeResponse.self)
        switch result {
        case .ok(let me):
            identity.save(me)
            identityIsKept = false
            state = .signedIn(me)
        case .sessionEnded:
            // The token died between adoption and /me (or was stale on restore).
            forgetIdentity()
            state = .signedOut
            if !quietOnFailure { errorMessage = "The session ended before it began — try again." }
        case .failed(_, let status) where quietOnFailure && (status == nil || status! >= 500):
            // A cold start with no signal, or a server having a bad minute, is not a dead session:
            // keep the tokens, so a saved event pass opened at a door with no bars does not sign
            // the person out behind it — and be the person kept on the phone, so tonight's door
            // and the programme are still theirs. The next restore tries the server again.
            keepGoingOffline()
        case .failed(let reason, _):
            // A sign-in that can't resolve /me is reported on the form, not
            // as the session-ended interrupt.
            expectingDeliberateEnd = true
            await tokens.endSession()
            forgetIdentity()
            state = .signedOut
            if !quietOnFailure { errorMessage = reason ?? "The server couldn't be reached." }
        case .rateLimited(let after):
            if quietOnFailure {
                keepGoingOffline()
            } else {
                state = .signedOut
                retryAfter = after ?? 60
            }
        }
    }

    /// The server couldn't be asked: stay the person kept on the phone if there is one, anonymous otherwise.
    private func keepGoingOffline() {
        if let kept = identity.load() {
            identityIsKept = true
            state = .signedIn(kept)
        } else {
            state = .signedOut
        }
    }

    private func forgetIdentity() {
        identity.clear()
        identityIsKept = false
    }

    private func handleSessionEnded() {
        forgetIdentity()
        if expectingDeliberateEnd {
            expectingDeliberateEnd = false
            return
        }
        state = .signedOut
        sessionEndedBanner = true
    }

    private func clearPending() {
        pendingEmail = nil
        pendingPassword = nil
    }
}

File: Ben.Data.WebApi.Client/Auth/SessionStore.cs
using Ben.Web.Services.WebApi;

namespace Ben.Data.WebApi.Client.Auth;

/// <summary>
/// The sign-in state machine every client screen reads:
/// signed out, authenticating, two-factor challenge, fetching identity, signed in.
/// </summary>
/// <remarks>
/// <para>A port of BenKit's <c>SessionStore</c>. Three decisions in here are load-bearing.</para>
///
/// <para><b>Roles come from the server, always.</b> Identity issues opaque data-protected tokens,
/// not JWTs, so there is nothing in an access token to decode. A client that tries gets nothing
/// back and quietly concludes the person is an ordinary member — which is how an administrator
/// ends up unable to see their own tools. Every path that acquires a token asks
/// <c>GET api/me</c> before reporting success.</para>
///
/// <para><b>A session ending is an interrupt, not a wall.</b> The store returns to signed-out and
/// raises a flag a screen can show; it does not tear the app down. Anonymous surfaces keep
/// working, because most of this product is readable without an account.</para>
///
/// <para><b>Restoring at cold start is quiet.</b> Stored tokens that the server no longer accepts
/// are an ordinary consequence of time passing, so the person is simply signed out. Announcing it
/// as a failure trains people to ignore the message that matters.</para>
/// </remarks>
public sealed class SessionStore
{
    private readonly TokenSession _session;
    private readonly IWebApiIdentityClient _identity;
    private readonly Func<CancellationToken, Task<ItemResult<MeResponse>>> _fetchMe;

    // Held only between the password step and the code step. A second factor is submitted to the
    // same /login endpoint as the password, so the credentials have to survive the round trip;
    // they are dropped the moment the challenge resolves either way.
    private string? _pendingEmail;
    private string? _pendingPassword;

    private bool _expectingDeliberateEnd;

    public SessionStore(
        TokenSession session,
        IWebApiIdentityClient identity,
        Func<CancellationToken, Task<ItemResult<MeResponse>>> fetchMe)
    {
        _session = session;
        _identity = identity;
        _fetchMe = fetchMe;
        _session.SessionChanged += OnSessionChanged;
    }

    /// <summary>The current state. Never null.</summary>
    public SessionState State { get; private set; } = SessionState.SignedOut;

    /// <summary>Fires whenever <see cref="State"/> changes.</summary>
    public event Action<SessionState>? StateChanged;

    /// <summary>
    /// Picks up a session stored by a previous run and confirms it is still good.
    /// </summary>
    /// <remarks>
    /// Optimistic then verified: the tokens are adopted first so an authenticated request can be
    /// made at all, and <c>GET api/me</c> settles whether they still work. A refusal lands quietly
    /// in signed-out with no banner — see the remarks on this class.
    /// </remarks>
    public async Task RestoreAsync(CancellationToken token = default)
    {
        await _session.RestoreAsync(token);
        if (!_session.HasSession)
        {
            Set(SessionState.SignedOut);
            return;
        }

        Set(new SessionState(SessionPhase.FetchingIdentity));

        var me = await _fetchMe(token);
        if (me.Failed || me.Item is null)
        {
            _expectingDeliberateEnd = true;   // an expired stored session is not news
            await _session.SignOutAsync();
            Set(SessionState.SignedOut);
            return;
        }

        Set(Resolved(me.Item));
    }

    /// <summary>
    /// Where a resolved identity actually lands.
    /// </summary>
    /// <remarks>
    /// An empty UserId is not a signed-in person. It is the server saying the token is valid but
    /// nothing here belongs to it — which only an external provider can produce. Reporting it as
    /// signed in would hand somebody an app in which every single page refuses them.
    /// </remarks>
    private static SessionState Resolved(MeResponse me) =>
        me.UserId == Guid.Empty
            ? new SessionState(SessionPhase.NeedsLocalAccount, me)
            : new SessionState(SessionPhase.SignedIn, me);

    /// <summary>Signs in with an email address and password.</summary>
    public Task SignInAsync(string email, string password, CancellationToken token = default)
        => AttemptAsync(email, password, twoFactorCode: null, recoveryCode: null, token);

    /// <summary>
    /// Answers a two-factor challenge, with either an authenticator code or a printed recovery code.
    /// </summary>
    /// <param name="isRecoveryCode">
    /// Which of the two the person typed. They go in different fields and Identity treats them
    /// differently; guessing by shape would spend a recovery code on a mistyped app code.
    /// </param>
    public Task SubmitTwoFactorAsync(string code, bool isRecoveryCode = false, CancellationToken token = default)
    {
        if (_pendingEmail is null || _pendingPassword is null)
        {
            // The challenge was abandoned or never started. Sending a code with no password would
            // be refused as bad credentials, which reads to the person as "your code was wrong".
            Set(SessionState.SignedOut);
            return Task.CompletedTask;
        }

        // Spaces and hyphens are how these are printed and how people read them aloud. The server
        // strips them too; rejecting a code typed comfortably is a self-inflicted failure.
        var cleaned = code.Replace(" ", string.Empty).Replace("-", string.Empty);

        return AttemptAsync(
            _pendingEmail, _pendingPassword,
            twoFactorCode: isRecoveryCode ? null : cleaned,
            recoveryCode: isRecoveryCode ? cleaned : null,
            token);
    }

    /// <summary>Abandons a two-factor challenge and forgets the held credentials.</summary>
    public void CancelTwoFactor()
    {
        ForgetPendingCredentials();
        Set(SessionState.SignedOut);
    }

    private async Task AttemptAsync(
        string email, string password,
        string? twoFactorCode, string? recoveryCode,
        CancellationToken token)
    {
        Set(new SessionState(SessionPhase.Authenticating));

        var attempt = await _identity.TryLoginAsync(email, password, twoFactorCode, recoveryCode, token);
        var response = attempt.Token;

        if (response is null || string.IsNullOrWhiteSpace(response.AccessToken))
        {
            var failure = LoginFailureMapping.From(attempt);

            if (failure == LoginFailure.RequiresTwoFactor)
            {
                _pendingEmail = email;
                _pendingPassword = password;
                Set(new SessionState(SessionPhase.TwoFactorChallenge, Failure: failure));
                return;
            }

            ForgetPendingCredentials();
            Set(new SessionState(
                SessionPhase.SignedOut,
                Failure: failure,
                RetryAfter: failure == LoginFailure.RateLimited ? attempt.RetryAfter : null));
            return;
        }

        ForgetPendingCredentials();
        await AdoptAsync(response, token);
    }

    /// <summary>
    /// Takes on a session that came from somewhere other than the password form — Sign in with
    /// Apple. The body that endpoint answers is the same one <c>/login</c> answers, deliberately,
    /// so there is nothing special to do with it.
    /// </summary>
    /// <remarks>
    /// NOT the Microsoft path. A Microsoft sign-in leaves the client holding a token issued by
    /// Microsoft rather than by us, so it is adopted directly into the token session with its own
    /// renewal and then settled by <see cref="ResolveIdentityAsync"/>.
    /// </remarks>
    public Task AdoptExternalSignInAsync(WebApiTokenResponse response, CancellationToken token = default)
        => AdoptAsync(response, token);

    /// <summary>
    /// Asks the server who the current token belongs to, and moves to the state that implies.
    /// </summary>
    /// <remarks>
    /// <para>Call after adopting a token this store did not itself obtain, and again after an
    /// external identity has been given a local account — the answer changes from an empty UserId
    /// to a real one, and nothing else tells the client that happened.</para>
    ///
    /// <para>A refusal here ends the session rather than leaving somebody in front of an app whose
    /// every page will refuse them.</para>
    /// </remarks>
    public async Task ResolveIdentityAsync(CancellationToken token = default)
    {
        Set(new SessionState(SessionPhase.FetchingIdentity));

        var me = await _fetchMe(token);
        if (me.Failed || me.Item is null)
        {
            _expectingDeliberateEnd = true;
            await _session.SignOutAsync();
            Set(new SessionState(SessionPhase.SignedOut, Failure: LoginFailure.UnknownRefusal));
            return;
        }

        Set(Resolved(me.Item));
    }

    private async Task AdoptAsync(WebApiTokenResponse response, CancellationToken token)
    {
        await _session.AdoptAsync(response, token);
        Set(new SessionState(SessionPhase.FetchingIdentity));

        var me = await _fetchMe(token);
        if (me.Failed || me.Item is null)
        {
            // Tokens that were accepted a moment ago and refused now. Rare, and worth naming
            // rather than reporting as a bad password: the credentials were plainly fine.
            _expectingDeliberateEnd = true;
            await _session.SignOutAsync();
            Set(new SessionState(SessionPhase.SignedOut, Failure: LoginFailure.UnknownRefusal));
            return;
        }

        Set(Resolved(me.Item));
    }

    /// <summary>Signs out at this person's request. There is no server endpoint; the tokens are simply dropped.</summary>
    public async Task SignOutAsync()
    {
        ForgetPendingCredentials();
        _expectingDeliberateEnd = true;
        await _session.SignOutAsync();
        Set(SessionState.SignedOut);
    }

    /// <summary>Clears the "your session ended" flag once a screen has shown it.</summary>
    public void AcknowledgeSessionEnded()
    {
        if (State.SessionEndedUnexpectedly)
            Set(State with { SessionEndedUnexpectedly = false });
    }

    private void OnSessionChanged(SessionEvent e)
    {
        if (e != SessionEvent.SessionEnded) return;

        var deliberate = _expectingDeliberateEnd;
        _expectingDeliberateEnd = false;

        ForgetPendingCredentials();
        Set(new SessionState(SessionPhase.SignedOut, SessionEndedUnexpectedly: !deliberate));
    }

    private void ForgetPendingCredentials()
    {
        _pendingEmail = null;
        _pendingPassword = null;
    }

    private void Set(SessionState state)
    {
        State = state;
        StateChanged?.Invoke(state);
    }
}
